fair/src/autogluon/fair/learners/fair.py

```diff
@@ -0,0 +1,418 @@
+        return out
+
+    def evaluate_predictor_regression(self,labels,y_pred, groups):
+        """Compute standard per-group  metrics for regression
+        """
+        assert False
+        #Dead code pathway needs rewritting 
+        base_metrics= np.stack((metrics.per_group.expected_label(labels, y_pred, groups),
+                                metrics.per_group.expected_prediction(labels, y_pred, groups)))
+        r=(base_metrics[0]-base_metrics[1]).reshape(1,-1)
+        scores=np.stack((metrics.per_group.r2                     (labels, y_pred, groups),
+                         metrics.per_group.mean_squared_error      (labels, y_pred, groups),
+                         metrics.per_group.root_mean_squared_error (labels, y_pred, groups),
+                         metrics.per_group.mean_absolute_error     (labels, y_pred, groups),
+                         metrics.per_group.median_absolute_error   (labels, y_pred, groups),
+                         metrics.per_group.spearmanr               (labels, y_pred, groups),
+                         metrics.per_group.pearsonr                (labels, y_pred, groups)))
+        scores=np.vstack ((base_metrics,r,scores))
+        gap=(scores.max(-1)-scores.min(-1)).reshape(-1,1)
+        names=('Expected label','Expected Prediction','r','r**2', 'RMS', 'mean AE',
+                 'median AE', 'Spearman r', 'pearson r'
+                 )
+        #print(collect,gap)
+        collect=np.hstack((scores,gap))
+        out=pd.DataFrame(collect.T,index =groups.unique().tolist()+ ["Maximum difference",],columns=names)
+        return out
+
+    def predict_proba(self,test_data):
+        "duplicates the functionality of classifier.predict_proba but with the new fair classifier"
+        proba=self.classifier.predict_proba(test_data)
+    
+        if isinstance(self.groups,str):
+            if self.infered_groups is False:
+                groups=np.asarray(test_data[self.groups])
+                
+        if self.infered_groups is False:
+            onehot=self.group_encoder.transform(groups.reshape(-1, 1)).toarray()
+        else:
+            onehot=np.asarray(self.infered_groups.predict_proba(test_data))
+        tmp=onehot.dot(self.offset)
+        proba+=tmp
+        return proba
+
+    def predict(self,test_data):
+        "duplicates the functionality of classifier.predict but with the new fair classifier"
+        proba=self.predict_proba(test_data)
+        return self.classifier.labels[proba.argmax(1)]
+
+
+def _needs_groups(fn):
+    "Internal helper function. Check if a metric requires a group argument or not"
+    #basically nothing works as some functions may take an implicit self or multiple keywords
+    #instead we just try to pass it two args and if it errors out return true
+    junk=np.zeros(1)
+    try:
+        fn(junk, junk)
+        return False
+    except Exception as e:
+        return True
```

Comment:
please no.

Add logic that adds an ability to check if an attribute / method exists in the metric that communicates if the metric needs a group argument. This is **not** the way to do that.

```if isinstance(fn, MyMetricClassThatRequiresGroup)```